Feat: 장바구니에 상품 추가 API 구현

GGHDMS opened the pull request:
변경사항
-----

장바구니에 상품 추가 api 구현
cart 도메인을 order 로 부터 분리

개발하면서 발생한 문제점(에러, 어려움)
-----

해결방안 또는 새롭게 알게된 점
-----
스크린샷
-----
당부하고 싶은 말 또는 논의해봐야할 점
-----

하나의 도메인 서비스에서 다른 도메인의 레포지토리를 참조하는 것 외의 방법을 찾아보는게 좋을 듯함

This closes #26

GGHDMS committed:
refactor: cart 도메인 을 따로 분리함

추가적으로 id 는 자동 생성이므로 builder 에서 받을 필요가 없다.

GGHDMS committed:
feat: 장바구니에 상품 추가하는 로직 구현

GGHDMS committed:
test: 기본적인 상품 추가 서비스 테스트 작성

sunwon12 commented on src/main/java/potato/potatoAPIserver/cart/domain/CartProduct.java:
```diff
@@ -28,8 +28,7 @@ public class CartProduct extends BaseTimeEntity {
     @Builder
-    public CartProduct(Long id, Cart cart, Product product, int quantity) {
-        this.id = id;
+    public CartProduct(Cart cart, Product product, int quantity) {
```
지금 ERD구조가 사용자의 장바구니가 있고 상품을 장바구니에 담기 위해서 CartProduct를 담는 구조인가?  Cart엔티티에 User속성뿐 아니라 Product도 같이 담으면 CartProduct는 필요 없을 것 같아. 그리고 현재 구조라면 필요없는 Join을 쓸 것 같아. 예를 들어 장바구니 상품을 찾는 SQL문은 유저 id로 Cart와 Join해서 UserId와 일치하는 Cart를 찾은 후 CartProduct에서 prouct를 찾잖아. Cart 엔티티에  User와 Product가 같이 있으면 해당하는 유저를 찾기 위해 Cart와 조인을 안 해도 되는거지

GGHDMS replied:
erd 를 그릴 때 고민하긴 했었는데 이런식으로 하면 유저가 카트를 여러개 가지고 있는 그림이 되서, 카트를 하나를 만들고 그 카트에 상품을 넣는 걸로 하긴했는데 개발하기 편한거는 너가 말한게 맞긴하지.. 바꾸는게 좋을까?

GGHDMS replied:
할인이랑 쿠폰 등이 생기면 카트를 따로 두는게 더 좋을 수도 있을거 같은데 어떻게 생각해?

sunwon12 replied:
내가 할인쿠폰을 개발해본 적이 없어서 이해가 잘 안 가능데 예시를 들어서 설명해줬어?

GGHDMS replied:
cart 를 따로 두면 totalprice, totaldiscountprice 등을 따로 만들수가 있어 확장성이 좋지 않을까 싶어서 말하거야

GGHDMS replied:
수량이랑 전체 가격을 조회 할 때마다 구하지 않아도 되서

sunwon12 replied:
장바구니에서 상품에 할인쿠폰을 적용한 가격을 보여주면 사용자 입장에서 편리하겠네! 그런데 만약에 할인 쿠폰이 생긴다면 그 때 쿠폰엔티티를 따로 만드는 게 좋을 것 같아. 현재는 장바구니에서 할인된 상품가격을 보여주는 요구사항이 없으니까 cart엔티티 하나로 만드는 게 좋을 것 같아.

GGHDMS replied:
서비스를 분리해서 수정했어 한번 확인해줘

sunwon12 replied:
웅 확인했어. Cart엔티티랑 cartproduct 분리한건 유지하는 거야?

GGHDMS replied:
cart 를 하나로 만들고 합친다 하더라도 유저의 카트를 찾는 부분만 없어질 뿐 다른 부분의 join 등은 결국 같아서 지금 상태가 더 괜찮은거 같아 나중에 확장성 까지 고려해서

sunwon12 replied:
알았우 ~

GGHDMS committed:
feat: 장바구니에 상품 추가하는 로직 구현

서비스 수정에 따른 컨트롤러 수정
`Min(1)` 을 `@Positive` 로 수정
entity 에서 수량을 직접 수정을 할 있게 코드 추가
서비스를 `CartService`, `CartProductService`로 분리

GGHDMS committed:
test: 서비스 로직 분리에 따른 테스트 수정

GGHDMS committed:
feat: 조회 기능은 readOnly = True 로 성능 향샹
